Mathlib/Topology/Compactification/OnePoint/ProjectiveLine.lean

```diff
@@ -68,4 +74,302 @@ lemma equivProjectivization_symm_apply_mk (x y : K) (h : (x, y) ≠ 0) :
     (equivProjectivization K).symm (mk K ⟨x, y⟩ h) = if y = 0 then ∞ else y⁻¹ * x := by
   simp [equivProjectivization]
 
+end DivisionRing
+
+section Field
+
+variable {K : Type*} [Field K] [DecidableEq K]
+
+/-- For a field `K`, the group `GL(2, K)` acts on `OnePoint K`. -/
+instance instGLAction : MulAction (GL (Fin 2) K) (OnePoint K) :=
+  (equivProjectivization K).mulAction (GL (Fin 2) K)
+
+lemma smul_infty_def {g : GL (Fin 2) K} :
+    g • ∞ = (equivProjectivization K).symm (.mk K (g 0 0, g 1 0) (fun h ↦ by
+      simpa [det_fin_two, Prod.mk_eq_zero.mp h] using g.det_ne_zero)) := by
+  simp [Equiv.smul_def, MulAction.compHom_smul_def, Projectivization.smul_mk, mulVec_eq_sum,
+    Units.smul_def]
+
+@[simp]
+lemma smul_infty_eq_ite {g : GL (Fin 2) K} :
+    (g • ∞ : OnePoint K) = if g 1 0 = 0 then ∞ else g 0 0 / g 1 0 := by
+  by_cases h : g 1 0 = 0 <;>
+  simp [h, div_eq_inv_mul, smul_infty_def]
+
+lemma smul_infty_eq_self_iff {g : GL (Fin 2) K} :
+    g • (∞ : OnePoint K) = ∞ ↔ g 1 0 = 0 := by
+  simp
+
+lemma smul_some_eq_ite {g : GL (Fin 2) K} {k : K} :
+    g • (k : OnePoint K) =
+      if g 1 0 * k + g 1 1 = 0 then ∞ else (g 0 0 * k + g 0 1) / (g 1 0 * k + g 1 1) := by
+  simp [Equiv.smul_def, MulAction.compHom_smul_def, Projectivization.smul_mk, mulVec_eq_sum,
+    div_eq_inv_mul, mul_comm, Units.smul_def]
+
+lemma map_smul {L : Type*} [Field L] [DecidableEq L]
+    (f : K →+* L) (g : GL (Fin 2) K) (c : OnePoint K) :
+    OnePoint.map f (g • c) = (g.map f) • (c.map f) := by
+  cases c with
+  | infty => simp [smul_infty_eq_ite, apply_ite]
+  | coe c => simp [smul_some_eq_ite, ← map_mul, ← map_add, apply_ite]
+
+end Field
+
 end OnePoint
+
+namespace Matrix
+
+variable {R n : Type*} [CommRing R] [Fintype n] [DecidableEq n]
+  (m : Matrix (Fin 2) (Fin 2) R) (g : GL (Fin 2) R)
+
+/-- The discriminant of a `2 × 2` matrix. -/
+noncomputable def disc : R := m.trace ^ 2 - 4 * m.det
+
+/-- A `2 × 2` matrix is *parabolic* if it is non-scalar and its discriminant is 0. -/
+def IsParabolic : Prop := m ∉ Set.range (scalar _) ∧ m.disc = 0
+
+/-- A `2 × 2` matrix is *hyperbolic* if its discriminant is a non-zero square. -/
+def IsHyperbolic : Prop := m.disc ≠ 0 ∧ ∃ s, s ^ 2 = m.disc
+
+/-- A `2 × 2` matrix is *elliptic* if its discriminant is not a square. -/
+def IsElliptic : Prop := ¬∃ s, s ^ 2 = m.disc
```

Comment:
Do we also want loxodromic? https://en.wikipedia.org/wiki/M%C3%B6bius_transformation#Loxodromic_transforms